File: app/src/main/java/com/mxj/mmitest/ui/testitems/ChargingTestActivity.java
package com.mxj.mmitest.ui.testitems;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.os.BatteryManager;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import com.mxj.mmitest.data.repository.TestRepository;
import com.mxj.mmitest.ui.base.BaseTestActivity;

/**
 * 充电测试 - 显示电池电量、状态、电压、温度、健康状态、充电方式
 */
public class ChargingTestActivity extends BaseTestActivity {

    private TestRepository repository;

    private LinearLayout mContainer;
    private TextView mLevelText;
    private TextView mStatusText;
    private TextView mVoltageText;
    private TextView mTemperatureText;
    private TextView mHealthText;
    private ProgressBar mBatteryProgress;
    private View mBatteryIcon;
    private BroadcastReceiver mReceiver;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setupContentView();
        super.onCreate(savedInstanceState);
        repository = TestRepository.getInstance(this);
    }

    private void setupContentView() {
        mContainer = new LinearLayout(this);
        mContainer.setOrientation(LinearLayout.VERTICAL);
        mContainer.setGravity(Gravity.CENTER);
        mContainer.setPadding(40, 60, 40, 40);

        // 标题
        TextView titleText = new TextView(this);
        titleText.setText("Battery & Charging Test");
        titleText.setTextSize(22);
        titleText.setTextColor(Color.parseColor("#333333"));
        titleText.setGravity(Gravity.CENTER);
        titleText.setPadding(0, 0, 0, 40);
        mContainer.addView(titleText);

        // 电池图标区域
        LinearLayout batteryIconLayout = new LinearLayout(this);
        batteryIconLayout.setOrientation(LinearLayout.VERTICAL);
        batteryIconLayout.setGravity(Gravity.CENTER);
        batteryIconLayout.setPadding(0, 0, 0, 30);

        // 电池外壳（外框）
        View batteryOutline = new View(this);
        batteryOutline.setBackgroundColor(Color.parseColor("#666666"));
        LinearLayout.LayoutParams outlineParams = new LinearLayout.LayoutParams(120, 60);
        outlineParams.setMargins(0, 0, 0, 5);
        batteryOutline.setLayoutParams(outlineParams);
        batteryIconLayout.addView(batteryOutline);

        // 电池电量填充
        mBatteryIcon = new View(this);
        mBatteryIcon.setBackgroundColor(Color.parseColor("#4CAF50"));
        LinearLayout.LayoutParams fillParams = new LinearLayout.LayoutParams(0, 50);
        fillParams.setMargins(10, 5, 10, 5);
        mBatteryIcon.setLayoutParams(fillParams);
        batteryIconLayout.addView(mBatteryIcon);

        mContainer.addView(batteryIconLayout);

        // 电量进度条
        mBatteryProgress = new ProgressBar(this, null, android.R.attr.progressBarStyleHorizontal);
        mBatteryProgress.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, 40));
        mBatteryProgress.setMax(100);
        mBatteryProgress.setProgress(0);
        mBatteryProgress.getProgressDrawable().setColorFilter(Color.parseColor("#4CAF50"), PorterDuff.Mode.SRC_IN);
        mContainer.addView(mBatteryProgress);

        // 电量百分比文本
        mLevelText = new TextView(this);
        mLevelText.setText("Battery Level: 0%");
        mLevelText.setTextSize(20);
        mLevelText.setTextColor(Color.parseColor("#333333"));
        mLevelText.setGravity(Gravity.CENTER);
        mLevelText.setPadding(0, 15, 0, 10);
        mContainer.addView(mLevelText);

        // 创建信息容器
        LinearLayout infoContainer = new LinearLayout(this);
        infoContainer.setOrientation(LinearLayout.VERTICAL);
        infoContainer.setBackgroundColor(Color.parseColor("#F5F5F5"));
        infoContainer.setPadding(20, 20, 20, 20);
        LinearLayout.LayoutParams infoParams = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        infoParams.setMargins(0, 20, 0, 0);
        infoContainer.setLayoutParams(infoParams);

        // 状态信息
        mStatusText = createInfoRow(this, infoContainer, "Status", "Checking...");
        mVoltageText = createInfoRow(this, infoContainer, "Voltage", "Unknown");
        mTemperatureText = createInfoRow(this, infoContainer, "Temperature", "Unknown");
        mHealthText = createInfoRow(this, infoContainer, "Health", "Unknown");

        mContainer.addView(infoContainer);

        // 提示文本
        TextView hintText = new TextView(this);
        hintText.setText("Connect charger to test charging function");
        hintText.setTextSize(16);
        hintText.setTextColor(Color.parseColor("#666666"));
        hintText.setGravity(Gravity.CENTER);
        hintText.setPadding(0, 30, 0, 0);
        mContainer.addView(hintText);

        setCustomContentView(mContainer);
    }

    private TextView createInfoRow(Context context, LinearLayout container, String label, String value) {
        LinearLayout row = new LinearLayout(context);
        row.setOrientation(LinearLayout.HORIZONTAL);
        row.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));

        TextView labelView = new TextView(context);
        labelView.setText(label + ": ");
        labelView.setTextSize(16);
        labelView.setTextColor(Color.parseColor("#666666"));
        labelView.setLayoutParams(new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1));

        TextView valueView = new TextView(context);
        valueView.setText(value);
        valueView.setTextSize(16);
        valueView.setTextColor(Color.parseColor("#333333"));
        valueView.setLayoutParams(new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1));
        valueView.setGravity(Gravity.END);

        row.addView(labelView);
        row.addView(valueView);
        container.addView(row);

        // 添加分隔线
        View divider = new View(context);
        divider.setBackgroundColor(Color.parseColor("#E0E0E0"));
        divider.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, 1));
        container.addView(divider);

        return valueView;
    }

    @Override
    protected void onResume() {
        super.onResume();
        IntentFilter filter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
        mReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                updateBatteryStatus(intent);
            }
        };
        registerReceiver(mReceiver, filter);
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mReceiver != null) {
            unregisterReceiver(mReceiver);
            mReceiver = null;
        }
    }

    private void updateBatteryStatus(Intent intent) {
        // 获取电池电量
        int level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);

        // 获取充电状态
        int status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1);
        boolean isCharging = (status == BatteryManager.BATTERY_STATUS_CHARGING || status == BatteryManager.BATTERY_STATUS_FULL);
        String statusText = getStatusText(status);

        // 获取电压（单位：mV）
        int voltage = intent.getIntExtra(BatteryManager.EXTRA_VOLTAGE, -1);
        String voltageText = voltage > 0 ? String.format("%.2fV", voltage / 1000.0) : "Unknown";

        // 获取温度（单位：0.1°C）
        int temperature = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, -1);
        String temperatureText = temperature > 0 ? String.format("%.1f°C", temperature / 10.0) : "Unknown";

        // 获取健康状态
        int health = intent.getIntExtra(BatteryManager.EXTRA_HEALTH, -1);
        String healthText = getHealthText(health);

        // 获取充电方式
        int plugType = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1);
        String plugText = getPlugTypeText(plugType);

        // 更新UI
        updateUI(level, statusText, voltageText, temperatureText, healthText, plugText, isCharging);

        // 如果有充电器连接或电池已满，自动启用PASS按钮
        boolean hasPowerSource = (plugType == BatteryManager.BATTERY_PLUGGED_AC ||
                plugType == BatteryManager.BATTERY_PLUGGED_USB ||
                plugType == BatteryManager.BATTERY_PLUGGED_WIRELESS);
        boolean testPassed = hasPowerSource || status == BatteryManager.BATTERY_STATUS_FULL;
        setPassEnabled(testPassed);
    }

    private void updateUI(int level, String status, String voltage, String temperature, String health, String plugType, boolean isCharging) {
        // 安全检查
        if (mBatteryProgress == null || mBatteryIcon == null ||
                mLevelText == null || mStatusText == null ||
                mVoltageText == null || mTemperatureText == null || mHealthText == null) {
            return;
        }

        // 更新电量进度条
        mBatteryProgress.setProgress(level);

        // 更新电池图标宽度（根据电量百分比）
        LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) mBatteryIcon.getLayoutParams();
        if (params != null) {
            params.width = (int) (100 * level / 100.0);
            mBatteryIcon.setLayoutParams(params);
        }

        // 根据电量改变颜色
        int batteryColor;
        if (level > 50) {
            batteryColor = Color.parseColor("#4CAF50"); // 绿色
        } else if (level > 20) {
            batteryColor = Color.parseColor("#FF9800"); // 橙色
        } else {
            batteryColor = Color.parseColor("#F44336"); // 红色
        }
        mBatteryIcon.setBackgroundColor(batteryColor);
        mBatteryProgress.getProgressDrawable().setColorFilter(batteryColor, PorterDuff.Mode.SRC_IN);

        // 更新文本
        mLevelText.setText(String.format("Battery Level: %d%%", level));
        mStatusText.setText(String.format("%s%s", status, !plugType.isEmpty() ? " (" + plugType + ")" : ""));
        mVoltageText.setText(voltage);
        mTemperatureText.setText(temperature);
        mHealthText.setText(health);
    }

    private String getStatusText(int status) {
        switch (status) {
            case BatteryManager.BATTERY_STATUS_CHARGING:
                return "Charging";
            case BatteryManager.BATTERY_STATUS_DISCHARGING:
                return "Discharging";
            case BatteryManager.BATTERY_STATUS_FULL:
                return "Full";
            case BatteryManager.BATTERY_STATUS_NOT_CHARGING:
                return "Not Charging";
            default:
                return "Unknown";
        }
    }

    private String getHealthText(int health) {
        switch (health) {
            case BatteryManager.BATTERY_HEALTH_GOOD:
                return "Good";
            case BatteryManager.BATTERY_HEALTH_OVERHEAT:
                return "Overheat";
            case BatteryManager.BATTERY_HEALTH_DEAD:
                return "Dead";
            case BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE:
                return "Over Voltage";
            case BatteryManager.BATTERY_HEALTH_UNSPECIFIED_FAILURE:
                return "Unspecified Failure";
            case BatteryManager.BATTERY_HEALTH_COLD:
                return "Cold";
            default:
                return "Unknown";
        }
    }

    private String getPlugTypeText(int plugType) {
        switch (plugType) {
            case BatteryManager.BATTERY_PLUGGED_AC:
                return "AC Charger";
            case BatteryManager.BATTERY_PLUGGED_USB:
                return "USB";
            case BatteryManager.BATTERY_PLUGGED_WIRELESS:
                return "Wireless";
            default:
                return "";
        }
    }

    @Override
    protected String getTestName() {
        return "充电测试";
    }

    @Override
    protected String getTestDescription() {
        return "电池和充电测试\n\n连接充电器检查充电功能";
    }

    @Override
    protected int getTimeoutSeconds() {
        return 30;
    }

    @Override
    protected String[] getRequiredPermissions() {
        return null;
    }

    @Override
    protected void onTestExecute() {
        // 测试执行
    }

    @Override
    protected boolean isPassEnabled() {
        return false; // 由电池状态决定
    }

    @Override
    protected void finishAndSaveResult(boolean passed) {
        repository.saveSingleTestResultSync(9, "充电测试", passed, getDeviceUniqueId());
        super.finishAndSaveResult(passed);
    }
}

File: app/src/main/java/com/mxj/mmitest/ui/testitems/ChargingTestActivity.kt
package com.mxj.mmitest.ui.testitems

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.lifecycleScope
import com.mxj.mmitest.data.repository.TestRepository
import com.mxj.mmitest.ui.base.BaseActivity
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class ChargingTestActivity : BaseActivity() {
    private val testName = "充电测试"
    private val timeoutSeconds = 30
    private val testItemId = 9
    private lateinit var repository: TestRepository

    // 电池状态常量
    private val batteryStatusMap = mapOf(
        BatteryManager.BATTERY_STATUS_UNKNOWN to "未知",
        BatteryManager.BATTERY_STATUS_CHARGING to "充电中",
        BatteryManager.BATTERY_STATUS_DISCHARGING to "放电中",
        BatteryManager.BATTERY_STATUS_NOT_CHARGING to "未充电",
        BatteryManager.BATTERY_STATUS_FULL to "已充满"
    )

    // 电池健康状态常量
    private val batteryHealthMap = mapOf(
        BatteryManager.BATTERY_HEALTH_UNKNOWN to "未知",
        BatteryManager.BATTERY_HEALTH_GOOD to "良好",
        BatteryManager.BATTERY_HEALTH_OVERHEAT to "过热",
        BatteryManager.BATTERY_HEALTH_DEAD to "电池损坏",
        BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE to "电压过高",
        BatteryManager.BATTERY_HEALTH_COLD to "温度过低"
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        repository = TestRepository(this)
        setContent {
            var remainingSeconds by remember { mutableIntStateOf(timeoutSeconds) }
            var batteryLevel by remember { mutableIntStateOf(0) }
            var batteryVoltage by remember { mutableIntStateOf(0) }
            var batteryTemperature by remember { mutableFloatStateOf(0f) }
            var batteryStatus by remember { mutableIntStateOf(BatteryManager.BATTERY_STATUS_UNKNOWN) }
            var batteryHealth by remember { mutableIntStateOf(BatteryManager.BATTERY_HEALTH_UNKNOWN) }
            var pluggedType by remember { mutableIntStateOf(0) }
            var batteryTechnology by remember { mutableStateOf("未知") }
            var isCharging by remember { mutableStateOf(false) }
            var passEnabled by remember { mutableStateOf(false) }

            // 充电来源
            val chargingSource = when (pluggedType) {
                BatteryManager.BATTERY_PLUGGED_AC -> "交流电 (AC)"
                BatteryManager.BATTERY_PLUGGED_USB -> "USB"
                BatteryManager.BATTERY_PLUGGED_WIRELESS -> "无线充电"
                else -> "未连接"
            }

            // 状态颜色
            val statusColor = when {
                isCharging -> Color(0xFF4CAF50) // 绿色
                batteryLevel < 20 -> Color(0xFFF44336) // 红色
                else -> Color(0xFFFF9800) // 橙色
            }

            // 背景颜色
            val backgroundColor = MaterialTheme.colorScheme.surface

            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .background(backgroundColor)
                    .padding(16.dp),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                // 顶部标题
                Text(
                    text = testName,
                    style = MaterialTheme.typography.headlineLarge,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.padding(bottom = 8.dp)
                )

                // 倒计时
                Text(
                    text = "剩余时间: ${remainingSeconds}秒",
                    style = MaterialTheme.typography.bodyMedium,
                    color = if (remainingSeconds <= 5)
                        MaterialTheme.colorScheme.error
                    else
                        MaterialTheme.colorScheme.onSurface,
                    modifier = Modifier.padding(bottom = 24.dp)
                )

                // 电池电量圆形指示器
                Box(
                    modifier = Modifier
                        .size(160.dp)
                        .clip(CircleShape)
                        .background(
                            if (isCharging) Color(0xFF1B5E20).copy(alpha = 0.1f)
                            else Color(0xFF1B5E20).copy(alpha = 0.05f)
                        ),
                    contentAlignment = Alignment.Center
                ) {
                    // 外圈
                    CircularProgressIndicator(
                        progress = { batteryLevel / 100f },
                        modifier = Modifier.size(140.dp),
                        color = statusColor,
                        strokeWidth = 12.dp,
                        trackColor = Color.Gray.copy(alpha = 0.2f)
                    )
                    // 电量文字
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Text(
                            text = "$batteryLevel%",
                            fontSize = 36.sp,
                            fontWeight = FontWeight.Bold,
                            color = statusColor
                        )
                        if (isCharging) {
                            Text(
                                text = "⚡ 充电中",
                                fontSize = 12.sp,
                                color = Color(0xFF4CAF50)
                            )
                        }
                    }
                }

                Spacer(modifier = Modifier.height(24.dp))

                // 状态提示
                Surface(
                    modifier = Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(12.dp),
                    color = if (isCharging) Color(0xFF4CAF50).copy(alpha = 0.1f)
                    else Color(0xFFFF9800).copy(alpha = 0.1f)
                ) {
                    Text(
                        text = if (isCharging) "✓ 已检测到充电 - 可以点击PASS" else "请连接充电器",
                        modifier = Modifier.padding(16.dp),
                        fontSize = 16.sp,
                        fontWeight = FontWeight.Medium,
                        color = if (isCharging) Color(0xFF4CAF50) else Color(0xFFFF9800)
                    )
                }

                Spacer(modifier = Modifier.height(16.dp))

                // 详细信息卡片
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(16.dp),
                    colors = CardDefaults.cardColors(
                        containerColor = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.5f)
                    )
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text(
                            text = "电池详细信息",
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Bold,
                            color = MaterialTheme.colorScheme.primary,
                            modifier = Modifier.padding(bottom = 12.dp)
                        )

                        InfoRow("充电状态", batteryStatusMap[batteryStatus] ?: "未知")
                        InfoRow("充电来源", chargingSource)
                        InfoRow("电池电压", "${batteryVoltage}mV")
                        InfoRow("电池温度", "${String.format("%.1f", batteryTemperature / 10.0)}°C")
                        InfoRow("电池健康", batteryHealthMap[batteryHealth] ?: "未知")
                        InfoRow("电池技术", batteryTechnology)
                    }
                }

                Spacer(modifier = Modifier.weight(1f))

                // 底部按钮
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(bottom = 16.dp),
                    horizontalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    // FAIL按钮
                    Button(
                        onClick = { saveAndFinish(false) },
                        modifier = Modifier
                            .weight(1f)
                            .height(56.dp),
                        colors = ButtonDefaults.buttonColors(
                            containerColor = MaterialTheme.colorScheme.error
                        ),
                        shape = RoundedCornerShape(12.dp)
                    ) {
                        Text(
                            text = "FAIL",
                            fontSize = 18.sp,
                            fontWeight = FontWeight.Bold
                        )
                    }

                    // PASS按钮
                    Button(
                        onClick = { saveAndFinish(true) },
                        enabled = passEnabled,
                        modifier = Modifier
                            .weight(1f)
                            .height(56.dp),
                        colors = ButtonDefaults.buttonColors(
                            containerColor = Color(0xFF4CAF50),
                            disabledContainerColor = Color.Gray.copy(alpha = 0.3f)
                        ),
                        shape = RoundedCornerShape(12.dp)
                    ) {
                        Text(
                            text = "PASS",
                            fontSize = 18.sp,
                            fontWeight = FontWeight.Bold,
                            color = if (passEnabled) Color.White else Color.Gray
                        )
                    }
                }
            }

            LaunchedEffect(Unit) {
                while (remainingSeconds > 0) {
                    updateBatteryInfo(
                        batteryManager = getSystemService(Context.BATTERY_SERVICE) as BatteryManager,
                        onLevelChange = { level, _ -> batteryLevel = level },
                        onVoltageChange = { voltage -> batteryVoltage = voltage },
                        onTemperatureChange = { temp -> batteryTemperature = temp },
                        onStatusChange = { status -> batteryStatus = status },
                        onHealthChange = { health -> batteryHealth = health },
                        onPluggedChange = { plugged -> pluggedType = plugged },
                        onTechnologyChange = { tech -> batteryTechnology = tech },
                        onChargingChange = { charging -> isCharging = charging }
                    )
                    passEnabled = isCharging
                    delay(1000)
                    remainingSeconds--
                }
                saveAndFinish(false)
            }
        }
    }

    @Composable
    private fun InfoRow(label: String, value: String) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 4.dp),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(
                text = label,
                fontSize = 14.sp,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Text(
                text = value,
                fontSize = 14.sp,
                fontWeight = FontWeight.Medium,
                color = MaterialTheme.colorScheme.onSurface
            )
        }
    }

    private fun updateBatteryInfo(
        batteryManager: BatteryManager,
        onLevelChange: (Int, Int) -> Unit,
        onVoltageChange: (Int) -> Unit,
        onTemperatureChange: (Float) -> Unit,
        onStatusChange: (Int) -> Unit,
        onHealthChange: (Int) -> Unit,
        onPluggedChange: (Int) -> Unit,
        onTechnologyChange: (String) -> Unit,
        onChargingChange: (Boolean) -> Unit
    ) {
        val intentFilter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
        val batteryStatus = registerReceiver(null, intentFilter)

        batteryStatus?.let { intent ->
            val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
            val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, 100)
            val voltage = intent.getIntExtra(BatteryManager.EXTRA_VOLTAGE, 0)
            val temperature = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0)
            val status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, BatteryManager.BATTERY_STATUS_UNKNOWN)
            val health = intent.getIntExtra(BatteryManager.EXTRA_HEALTH, BatteryManager.BATTERY_HEALTH_UNKNOWN)
            val plugged = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, 0)
            val technology = intent.getStringExtra(BatteryManager.EXTRA_TECHNOLOGY) ?: "未知"

            onLevelChange(level, scale)
            onVoltageChange(voltage)
            onTemperatureChange(temperature.toFloat())
            onStatusChange(status)
            onHealthChange(health)
            onPluggedChange(plugged)
            onTechnologyChange(technology)
            onChargingChange(status == BatteryManager.BATTERY_STATUS_CHARGING ||
                    status == BatteryManager.BATTERY_STATUS_FULL)
        }
    }

    private fun saveAndFinish(passed: Boolean) {
        lifecycleScope.launch {
            repository.saveSingleTestResult(
                testItemId = testItemId,
                testItemName = testName,
                passed = passed,
                deviceId = android.provider.Settings.Secure.getString(
                    contentResolver,
                    android.provider.Settings.Secure.ANDROID_ID
                ) ?: android.os.Build.MODEL
            )
            finish()
        }
    }
}
